src/procedural-macros.md

```diff
@@ -1,24 +1,546 @@
+    let mut b = iter.next().unwrap();
+    let a_span = a.span();
+    a.set_span(b.span());
+    b.set_span(a_span);
+    return vec![a, b].into_iter().chain(iter).collect()
+}
+```
+
+We can see what's going on here by feeding invalid syntax into the macro and
+seeing what the compiler reports. Let's start off by seeing what the compiler
+does normally:
+
+```rust,ignore
+// src/main.rs
+fn _() {}
+```
+
+is compiled as:
+
+```sh
+$ cargo run
+   Compiling foo v0.1.0 (file://.../foo)
+error: expected identifier, found reserved identifier `_`
+ --> src/main.rs:1:4
+  |
+1 | fn _() {}
+  |    ^ expected identifier, found reserved identifier
+
+error: aborting due to previous error
+```
+
+but when we feed it through our macro:
+
+```rust,ignore
+extern crate my_macro;
+
+my_macro::swap_spans! {
+    fn _() {}
+}
+
+fn main() {}
+```
+
+and compile it:
+
+```sh
+$ cargo run
+   Compiling foo v0.1.0 (file://.../foo)
+error: expected identifier, found reserved identifier `_`
+ --> src/main.rs:4:5
+  |
+4 |     fn _() {}
+  |     ^^ expected identifier, found reserved identifier
+
+error: aborting due to previous error
+```
+
+notice how the error message is pointing to the wrong span! This is because we
+swapped the spans of the first two tokens, giving the compiler false information
+about where the tokens came from.
+
+Controlling spans is quite a powerful feature and needs to be used with care,
+misuse can lead to some excessively confusing error messages!
+
+### Procedural macros and hygiene
+
+Currently all procedural macros are "unhygienic". This means that all procedural
+macros behave as if the output token stream was simply written inline to the
+code it's next to. This means that it's affected by external items and also
+affected by external imports and such.
+
+Macro authors need to be careful to ensure their macros work in as many contexts
+as possible given this limitation. This often includes using absolute paths to
+items in libraries (`::std::option::Option` instead of `Option`) or
+by ensuring that generated functions have names that are unlikely to clash with
+other functions (like `__internal_foo` instead of `foo`).
+
+Eventually more support for hygiene will be added to make it easier to write a
+robust macro.
+
+### Limitations of procedural macros
+
+Procedural macros are not currently quite as powerful as `macro_rules!`-defined
+macros in some respects. These limitations include:
+
+* Procedural bang macros can only be invoked in *item* contexts. For example you
+  can't define your own `format!` yet because that's only ever invoked in an
+  expression context. Put another way, procedural bang macros can only expand to
+  items (things like functions, impls, etc), not expressions.
+
+  This is primarily due to the lack of hygiene with procedural macros. Once a
+  better hygiene story is stabilized this restriction will likely be lifted.
+
+* Procedural macros cannot expand to definitions of `macro_rules!` macros (none
```

Comment:
Hmm, serde_derive definitely does this now and it's stable. Maybe I misunderstand what this bullet means. Our two derives each include a macro_rules helper macro in the generated code that is called from elsewhere in the generated code.

Reply:
The restriction was implemented much later than derives were stabilized (https://github.com/rust-lang/rust/pull/50820), so it was applied to fn-like macros and attribute macros to avoid breakage.

So macro_rules generated from derives do not produce immediate errors, but they can still be buggy and perhaps go through some breakage later with hygiene changes.